feat(list): allow semantic tags (mainly, nav on List and a on ListItem)

4cm4k1 opened the pull request:
### Backstory
This PR began as an attempt to help with the focus-trap issues that using `Drawer` and `List` together were causing (reported here: https://github.com/material-components/material-components-web-react/issues/382; tracked here: https://github.com/material-components/material-components-web-react/issues/433).

### Purpose
What this PR now is simply adding the ability to override the default tags (`nav` [default: `ul`] on `List` and `a` [default: `li`] on `ListItem`) by using the `tag` prop.

### Context
Changing the tags would be recommended when using these components to render a page's navigation, as it is semantic HTML and helps with a11y. Additionally, this may help with compatibility with routing libraries such as [React Router](https://github.com/ReactTraining/react-router) and [Next.js' Link](https://github.com/zeit/next.js#with-link).

#### Notes
I've also added two unit tests corresponding to the functionality added by this PR and added text regarding this feature in `List`'s `README.md`.

moog16 commented on packages/list/ListItem.js:
```diff
@@ -72,14 +72,16 @@ export default class ListItem extends Component {
       ...otherProps
     } = this.props;
 
+    const SemanticListItem = this.props.href ? 'a' : 'li';
```
I do agree with this change, but I think there should be a way to programatically change routes. Things like React Router and SPA architecture require this. Semantically we will want an anchor tag, but no href, correct?

4cm4k1 replied:
I think so? I'm having a hard time with the layers of abstraction here. Below is an example of how I would imagine it working while using [`next`](https://github.com/zeit/next.js) and its routing logic via `next/link`.

```jsx
import Link from 'next/link';
import List, { 
  ListItem,
  ListItemGraphic,
  ListItemText,
} from '@material/react-list';

const NavList = [
  {
    href: '/',
    name: 'Home',
    icon: 'home',
  }, {
    href: '/info',
    name: 'Info',
    icon: 'info',
  }
];

export default () => (
  <List wrapFocus tag="nav">
    {NavList.map(({ href, name, icon }, index) => (
      <Link key={index} href={href} prefetch passHref>
        <ListItem key={index} tag="a">
          <ListItemGraphic graphic={<MaterialIcon icon={icon} />} />
          <ListItemText primaryText={name} />
        </ListItem>
      </Link>
      // With react-router-dom it would be pretty similar...
      // <Link to={href}>
      //   <ListItem ...>
      //     <ListItemGraphic ... />
      //     <ListItemText ... />
      //   </ListItem>
      // </Link>
    ))}
  </List>
);
```

moog16 replied:
What if people wanted to do 
```
<ListItem key={index} tag="a" onClick={this.goToAnotherRoute}>
          <ListItemGraphic graphic={<MaterialIcon icon={icon} />} />
          <ListItemText primaryText={name} />
</ListItem>
```

All I'm saying is, we should remove `const SemanticListItem = this.props.href ? 'a' : 'li';`, in favor of the `Tag` property.

4cm4k1 replied:
> What if people wanted to do
> 
> ```
> <ListItem key={index} tag="a" onClick={this.goToAnotherRoute}>
>           <ListItemGraphic graphic={<MaterialIcon icon={icon} />} />
>           <ListItemText primaryText={name} />
> </ListItem>
> ```
> All I'm saying is, we should remove `const SemanticListItem = this.props.href ? 'a' : 'li';`, in favor of the `Tag` property.

Got it. I've updated `ListItem` accordingly and updated the README.

Another thing to consider: When using router frameworks that nest elements within `<Link></Link>`, I encountered problems where the error output is something to the effect of `Cannot set property 'aria-selected' of undefined`, and when using the `selectedIndex` prop, it’s unable to set it on the index specified. Would this be fixed if users extended the `Link` component to pass the necessary props to its children (i.e., `ListItem`)? Or how else can we make it safe to nest `ListItem`?

I have the beginnings of that in [my project](https://github.com/4cm4k1/website/blob/master/components/Link.jsx), though currently using RMWC because of the crashing issues here, shown here:

```jsx
<Link {...props}>
  <ListItem tag="a" activated={activated}>
    {children}
  </ListItem>
</Link>
```

4cm4k1 commented:
@moog16 Thank you for pointing that out. `ListItemText`, `ListItemGraphic`, and `ListItemMeta` have `tabIndex` attributes currently, while `ListItem` does not. What would be the best way, then, to implement it for `ListItem`? I guess I'm a bit confused as to why the children of `ListItem` have `tabIndex` but not `ListItem`.

I updated the readme with the `tag` prop name addition to `List`.

moog16 commented:
@4cm4k1 tabIndex is rather complicated, but necessary for a11y. WCAG states we should be able to tab to each separate <List>, but arrow through each <ListItem>. We also must add tabindex=-1 to all child elements, since some of them might have tabIndex.

I haven't had time this week to check, but you're saying using <List> with tabIndex does not solve modal issue of crashing?

4cm4k1 commented:
> @4cm4k1 tabIndex is rather complicated, but necessary for a11y. WCAG states we should be able to tab to each separate , but arrow through each . We also must add tabindex=-1 to all child elements, since some of them might have tabIndex.
> 
> I haven't had time this week to check, but you're saying using with tabIndex does not solve modal issue of crashing?

Correct. The crash issue only seemed to go away after using `<a>` tags instead of `<li>`. I totally understand and support the need for a11y; I’m just admittedly not aware of the technical details.

moog16 commented:
@4cm4k1 could you please fix conflicts? I'll do another review after that is done. Thanks!

moog16 commented:
@4cm4k1 opened this issue to solve the modal crashing issue https://github.com/material-components/material-components-web-react/issues/433

4cm4k1 commented:
@4cm4k1 I resolved the merge conflicts, added some text to the `README.md`, renamed this PR, and rewrote the PR's description to better reflect what this is. Thanks! 😄

4cm4k1 commented on test/unit/list/ListItem.test.js:
```diff
@@ -65,3 +65,8 @@ test('passes props.childrenTabIndex to children as props.tabIndex', () => {
   );
   assert.equal(wrapper.find('.list-item-child').props().tabIndex, 2);
 });
+
```
Can you clarify how to get set up locally? I've been unable to successfully `npm run build` or `npm run test:unit`, which has made it difficult to validate these tests myself, and there doesn't seem to be any documentation on 'developing' the repo. For now, I've added the default cases and corrected a mistake in the list item `a` test. Thanks!

moog16 replied:
Sounds like you've got it figured out

moog16 replied:
Added a task to add a doc for contributions. Thanks! https://github.com/material-components/material-components-web-react/issues/440

moog16 commented:
created a tests pr https://github.com/material-components/material-components-web-react/pull/439

moog16 commented:
@4cm4k1 tests pass! Please sign it by commenting "I signed it" in this pr

4cm4k1 commented:
I signed it

moog16 commented:
Thanks @4cm4k1 !!
